### ISSUE TITLE
Expand all initially?

### ISSUE THREAD

**Issue by adamgruber** [NeekSandhu/react-aspen#2]
Is it possible to have the tree fully expanded initially? Can you suggest how to implement expand all / collapse all functionality?

**Comment by NeekSandhu** [comment]
# Preface

The reason such a feature is not available in the core is that the operation strategy varies depending on how "fast" `Host#getItems(path)` is. Therefore, no assumption was made.

Keep in mind that `getItems` and `expandDirectory/collapseDirectory` are deeply linked, in that, the latter calls `getItems` after all.

This is a succinct version of how the call stack will look like for a recursive `expandDirectory`:

```
expandRecursive                     app

getItems                            aspen
expandDirectory                     aspen
render                              react
↻
getItems                            aspen
expandDirectory                     aspen
render                              react
↻
getItems                            aspen
expandDirectory                     aspen
render                              react
↻
...
```

A recursive `collapseDirectory` will be the same except there won't be any call to `getItems`.

# Challenges

As you can see, there will be a lot of calls to `render`, if done the "typical" way, i.e start recursing top-down and keep calling `expandDirectory` on the way. This may cause a performance hit, although, it may not. More on that below.

## Case 1

If `getItems` is almost instantaneous, then you want to keep the number of "re-renders" to a minimum. Since visually speaking, the above call stack will be done in milliseconds, and to the user, it'd appear as if it was just one "re-paint", but really, you'll be calling `render` many-many times for no apparent reason. Wasting CPU cycles on something that won't yield a benefit (worse in case you have side-effects that run on each render).

Above case can be fixed by doing what I call, "backstage" mode or in other words, expanding/collapsing the entire tree "off-screen" and finishing it off with just one "render" pass. Explained below.

## Case 2

If `getItems` is a bit slow, i.e,  slow enough for a user to pick it as such. Then it's best to not do it the "backstage"/"off-screen" way, as keeping the user in the dark for too long might make them unconfident, so in this case, `render` what you get, as you get.

Above case is the "typical" way, as seen in the example call stack above.

# Solutions 

Case 2 is simple, so we'll only tackle Case 1, for when `getItems` is almost instantaneous, we want the call stack to look like this:

```diff
  expandRecursive                     app
+ begin                               app

  getItems                            aspen
  expandDirectory                     aspen
- render                              react
  ↻
  getItems                            aspen
  expandDirectory                     aspen
- render                              react
  ↻
  getItems                            aspen
  expandDirectory                     aspen
- render                              react
  ↻
  ...
+ finishOff                           app
+ render                              react
```

Again, same for `collapseDirectory` except that there won't be `getItems`.

Here the points of interest are `begin` and `finishOff`, which we need to implement and rest will take care of itself (if done right).

The idea is simple, any operation, expand or collapse, if executed on a directory which is child of **collapsed** directory will be **implicit**. Meaning, the effect will occur, it just won't show up until that said collapsed directory is expanded. The trick is in how aspen wires view and the model. Model will update no matter what, but the view will only update if that change is **visible** to the user, i.e, if it happens inside of an (already) expanded directory.

**tl;dr:** [`Root#isItemVisibleAtSurface(fileOrDir)`](https://neeksandhu.github.io/aspen-core/classes/root.html#isitemvisibleatsurface)
___

### Instructions

#### You'll need:

- [`Root#iterateTopDown(cb, startingPoing?)`](https://neeksandhu.github.io/aspen-core/classes/root.html#iteratetopdown)
- [`Root#expandDirectory(directory, ensureVisible? = true)`](https://neeksandhu.github.io/aspen-core/classes/root.html#expanddirectory)
- [`Root#collapseDirectory(directory)`](https://neeksandhu.github.io/aspen-core/classes/root.html#collapsedirectory)
- [`Directory#expanded`](https://neeksandhu.github.io/aspen-core/classes/directory.html#expanded)

#### Recursive expand

Start iterating from a `startingPoint` of your choice (default is `root` itself). Your callback will look like this:

```
(item: FileOrDir, stepOver: function, stepIn: function, stepOut: function, exit: function): void
```

[`TopDownIteratorCallback` documentation](https://neeksandhu.github.io/aspen-core/globals.html#topdowniteratorcallback)

Now, use a mix of `stepIn`s, `stepOut`s and `next`s until you find the first directory whose `Directory#expanded` is `false`. Then while you're inside the "outer" iterator (we'll come back to it later), call `Root#iterateTopDown` with the said directory as the `startingPoint`. For this, use a mix of `stepIn`s, `stepOut`s and `next`s again to reach directory and `Root#expandDirectory(dir, false)` on it. Remember to have the second argument `ensureVisible` as `false`, for now, we want to alter just the model, not the view.

Once the inner iterator is finished calling `expandDirectory(dir, false)` on each recursive child of the *still collapsed parent*, we're ready to exit inner iterator, come back to outer iterator and call `expandDirectory(dir, true)` on it. And boom, that entire tree is expanded in a single `render`.

In the outer iterator, find the `next` directory whose `expanded` is `false` and repeat the process.

> 💡
> You'll also need to leverage `stepOver` to jump over `FileEntry`s; `depth` to keep track of how deep you are in that iterator (relative to `startingPoint`). If `item` is `undefined` that means there are no more items at that `depth`, in that case you must `stepOut` > `next` > `stepIn`, or `exit` if `!item && depth === 0` 

#### Recursive collapse

Start the iterator like before at a `startingPoint` of your choice. Find first directory whose `expanded` is `true`, collapse it first (this will "disconnect" that node from view), then start an inner iterator with itself being as the `startingPoint`, call `collapseDirectory` on each child.

Repeat for each, in the outer iterator.

Like before, the entire (sub)tree will be collapsed at the model level, but only one repaint will be required as far as the view is concerned.

# Outro

I hope this helps, feel free to ask any more questions. Please consider publishing your implementation on `npm` if that's possible.

**Comment by NeekSandhu** [comment]
This may *look* complicated to implement, but trust me, it's not. Just the problem is a bit complex.

Aspen's core principles were, are and will always be top-notch performance!!

The selling point of Aspen is the fact that it definitely **does** recursion, just like any other implementation, except that it does it **differently**.

![](https://memegenerator.net/img/instances/81941927/the-hardest-choices-require-the-strongest-wills.jpg)

**Comment by adamgruber** [comment]
Thanks for the detailed explanation. Definitely going to try this out.

**Comment by NeekSandhu** [comment]
Closing this as stale. Please re-open if necessary :)
